File: views/rollView.py
import discord
from datetime import datetime
from utils.classGenerator import generer_classe, generer_arme_aleatoire
from utils.stats import update_stats
from functools import lru_cache

class RollView(discord.ui.View):
    def __init__(self, classe):
        super().__init__(timeout=300)
        self.classe = classe

    @discord.ui.button(label='🔄 RE-ROLL', style=discord.ButtonStyle.primary, row=0)
    async def reroll(self, interaction: discord.Interaction, button: discord.ui.Button):
        # Réponse immédiate pour montrer que le bouton a été cliqué
        await interaction.response.defer()
        
        update_stats("reroll")
        nouvelle_classe = generer_classe()
        self.classe = nouvelle_classe
        embed = create_class_embed(nouvelle_classe)
        
        # Édition après coup pour une meilleure réactivité
        await interaction.edit_original_response(embed=embed, view=self)

    @discord.ui.button(label='🎯 ARME SEULE', style=discord.ButtonStyle.success, row=0)
    async def arme_seule(self, interaction: discord.Interaction, button: discord.ui.Button):
        # Réponse immédiate
        await interaction.response.defer()
        
        # Import local pour éviter les imports circulaires
        from views.armeView import ArmeView
        arme = generer_arme_aleatoire()
        embed = discord.Embed(title="🎯 Arme Aléatoire", color=0x0099ff, timestamp=datetime.now())
        embed.add_field(name="🔫 Votre arme", value=f"```{arme}```", inline=False)
        embed.set_footer(text="🍀 Bonne chance !")
        
        await interaction.edit_original_response(embed=embed, view=ArmeView(self.classe))

    @discord.ui.button(label='🏆 DÉFI', style=discord.ButtonStyle.danger, row=0)
    async def defi(self, interaction: discord.Interaction, button: discord.ui.Button):
        # Réponse immédiate
        await interaction.response.defer()
        
        # Import local pour éviter les imports circulaires
        from views.defiView import DefiView
        from utils.classGenerator import defis_data
        view = DefiView(self.classe)
        embed = discord.Embed(
            title="🏆 Choisissez votre défi !",
            description="Sélectionnez le niveau de difficulté :",
            color=0xff4444,
            timestamp=datetime.now()
        )
        for niveau, defis_list in defis_data.items():
            if isinstance(defis_list, list):
                embed.add_field(
                    name=f"{niveau.title()}",
                    value=f"```{len(defis_list)} défis disponibles```",
                    inline=True
                )
        
        await interaction.edit_original_response(embed=embed, view=view)

    @discord.ui.button(label='🔫 ARME PRINCIPALE', style=discord.ButtonStyle.secondary, row=1)
    async def arme_principale_direct(self, interaction: discord.Interaction, button: discord.ui.Button):
        await interaction.response.defer()
        
        # Import local pour éviter les imports circulaires
        from views.principaleView import PrincipaleView
        update_stats("principale")
        
        view = PrincipaleView()
        embed = discord.Embed(
            title="🔫 Armes principales",
            description="```Choisissez une catégorie pour obtenir une arme aléatoire dedans.```",
            color=0x00ccff,
            timestamp=datetime.now()
        )
        await interaction.edit_original_response(embed=embed, view=view)

    @discord.ui.button(label='🗡️ ARME SECONDAIRE', style=discord.ButtonStyle.secondary, row=1)
    async def arme_secondaire_direct(self, interaction: discord.Interaction, button: discord.ui.Button):
        await interaction.response.defer()
        
        # Import local pour éviter les imports circulaires
        from views.secondaireView import SecondaireView
        update_stats("secondaire")
        
        view = SecondaireView()
        embed = discord.Embed(
            title="🗡️ Armes secondaires",
            description="```Choisissez une catégorie (Pistolets, Lanceurs ou Spécial) pour obtenir une arme aléatoire dedans.```",
            color=0x00ccff,
            timestamp=datetime.now()
        )
        await interaction.edit_original_response(embed=embed, view=view)

    @discord.ui.button(label='SAUVEGARDER', style=discord.ButtonStyle.success, emoji='💾', row=2)
    async def sauvegarder_classe(self, interaction: discord.Interaction, button: discord.ui.Button):
        await interaction.response.defer(ephemeral=True)
        
        try:
            # Trouver le salon "classe" dans le serveur
            salon_classe = None
            for channel in interaction.guild.text_channels:
                if channel.name == "🔫・classe":
                    salon_classe = channel
                    break
            
            if not salon_classe:
                return await interaction.followup.send(
                    "❌ **Salon '🔫・classe' introuvable** - Assurez-vous qu'un salon textuel nommé '🔫・classe' existe sur ce serveur.",
                    ephemeral=True
                )
            
            # Créer l'embed pour la classe sauvegardée
            embed = discord.Embed(
                title="💾 Classe Sauvegardée",
                description=f"**Sauvegardée par {interaction.user.mention}**",
                color=0x00ff00,
                timestamp=datetime.now()
            )
            embed.add_field(name="🔫 Arme n°1", value=f"```{self.classe['arme_principale']}```", inline=True)
            embed.add_field(name="🔫 Arme n°2", value=f"```{self.classe['arme_secondaire']}```", inline=True)
            embed.add_field(name="", value="", inline=True)  # Spacer
            embed.add_field(name="⚡ Atout 1", value=f"```{self.classe['atout_1']}```", inline=True)
            embed.add_field(name="⚡ Atout 2", value=f"```{self.classe['atout_2']}```", inline=True)
            embed.add_field(name="⚡ Atout 3", value=f"```{self.classe['atout_3']}```", inline=True)
            embed.add_field(name="🎯 Équipement tactique", value=f"```{self.classe['equipement_tactique']}```", inline=True)
            embed.add_field(name="💥 Équipement mortel", value=f"```{self.classe['equipement_mortel']}```", inline=True)
            embed.add_field(name="", value="", inline=True)  # Spacer
            embed.set_footer(text=f"Sauvegardée depuis #{interaction.channel.name}")
            
            # Envoyer la classe dans le salon "classe"
            await salon_classe.send(embed=embed)
            
        except Exception as e:
            await interaction.followup.send(
                f"❌ **Erreur lors de la sauvegarde :** {str(e)}",
                ephemeral=True
            )

@lru_cache(maxsize=128)  # Cache pour éviter de recréer les mêmes embeds
def _create_embed_cached(arme_principale, arme_secondaire, atout_1, atout_2, atout_3, equipement_tactique, equipement_mortel, timestamp_str):
    """Version cachée de create_class_embed pour de meilleures performances"""
    embed = discord.Embed(title="🎲 Classe Générée", color=0x0099ff, timestamp=datetime.fromisoformat(timestamp_str))
    embed.add_field(name="🔫 Arme n°1", value=f"```{arme_principale}```", inline=True)
    embed.add_field(name="🔫 Arme n°2", value=f"```{arme_secondaire}```", inline=True)
    embed.add_field(name="", value="", inline=True)  # Spacer
    embed.add_field(name="⚡ Atout 1", value=f"```{atout_1}```", inline=True)
    embed.add_field(name="⚡ Atout 2", value=f"```{atout_2}```", inline=True)
    embed.add_field(name="⚡ Atout 3", value=f"```{atout_3}```", inline=True)
    embed.add_field(name="🎯 Équipement tactique", value=f"```{equipement_tactique}```", inline=True)
    embed.add_field(name="💥 Équipement mortel", value=f"```{equipement_mortel}```", inline=True)
    embed.add_field(name="", value="", inline=True)  # Spacer
    embed.set_footer(text="🍀 Bonne chance!")
    return embed

def create_class_embed(classe):
    """Créer un embed Discord pour afficher la classe générée (version optimisée)"""
    # Utiliser la version cachée avec un timestamp arrondi à la seconde pour améliorer le cache hit rate
    timestamp_str = datetime.now().replace(microsecond=0).isoformat()
    
    return _create_embed_cached(
        classe['arme_principale'],
        classe['arme_secondaire'], 
        classe['atout_1'],
        classe['atout_2'],
        classe['atout_3'],
        classe['equipement_tactique'],
        classe['equipement_mortel'],
        timestamp_str
    )

File: views/rollView.js
import { ActionRowBuilder, ButtonBuilder, ButtonStyle, EmbedBuilder } from 'discord.js';
import { generateClass, generateRandomWeapon } from '../utils/classGenerator.js';
import { updateStats } from '../utils/stats.js';

/**
 * Create an embed for displaying a generated class
 * @param {Object} classe - The generated class object
 * @returns {EmbedBuilder} Discord embed
 */
export function createClassEmbed(classe) {
    const embed = new EmbedBuilder()
        .setTitle('🎲 Classe Générée')
        .setColor(0x0099ff)
        .setTimestamp()
        .addFields(
            { name: '🔫 Arme n°1', value: `\`\`\`${classe.arme_principale}\`\`\``, inline: true },
            { name: '🔫 Arme n°2', value: `\`\`\`${classe.arme_secondaire}\`\`\``, inline: true },
            { name: '', value: '', inline: true }, // Spacer
            { name: '⚡ Atout 1', value: `\`\`\`${classe.atout_1}\`\`\``, inline: true },
            { name: '⚡ Atout 2', value: `\`\`\`${classe.atout_2}\`\`\``, inline: true },
            { name: '⚡ Atout 3', value: `\`\`\`${classe.atout_3}\`\`\``, inline: true },
            { name: '🎯 Équipement tactique', value: `\`\`\`${classe.equipement_tactique}\`\`\``, inline: true },
            { name: '💥 Équipement mortel', value: `\`\`\`${classe.equipement_mortel}\`\`\``, inline: true },
            { name: '', value: '', inline: true } // Spacer
        )
        .setFooter({ text: '🍀 Bonne chance!' });
    
    return embed;
}

/**
 * Create the main roll view with all interaction buttons
 * @param {Object} classe - The initial generated class
 * @returns {Object} Components for the message
 */
export function createRollView(classe) {
    const row1 = new ActionRowBuilder()
        .addComponents(
            new ButtonBuilder()
                .setCustomId('reroll')
                .setLabel('🔄 RE-ROLL')
                .setStyle(ButtonStyle.Primary),
            new ButtonBuilder()
                .setCustomId('arme_seule')
                .setLabel('🎯 ARME SEULE')
                .setStyle(ButtonStyle.Success),
            new ButtonBuilder()
                .setCustomId('defi')
                .setLabel('🏆 DÉFI')
                .setStyle(ButtonStyle.Danger)
        );

    const row2 = new ActionRowBuilder()
        .addComponents(
            new ButtonBuilder()
                .setCustomId('arme_principale_direct')
                .setLabel('🔫 ARME PRINCIPALE')
                .setStyle(ButtonStyle.Secondary),
            new ButtonBuilder()
                .setCustomId('arme_secondaire_direct')
                .setLabel('🗡️ ARME SECONDAIRE')
                .setStyle(ButtonStyle.Secondary)
        );

    return { components: [row1, row2] };
}

/**
 * Handle button interactions for the roll view
 * @param {Interaction} interaction - Discord interaction
 * @param {Object} classe - Current class data
 * @returns {Object|null} New class data if changed, null otherwise
 */
export async function handleRollInteraction(interaction, classe) {
    const customId = interaction.customId;

    switch (customId) {
        case 'reroll':
            await interaction.deferUpdate();
            updateStats('reroll');
            const nouvelleClasse = generateClass();
            const embed = createClassEmbed(nouvelleClasse);
            const view = createRollView(nouvelleClasse);
            await interaction.editReply({ embeds: [embed], ...view });
            return nouvelleClasse;

        case 'arme_seule':
            await interaction.deferUpdate();
            const arme = generateRandomWeapon();
            const armeEmbed = new EmbedBuilder()
                .setTitle('🎯 Arme Aléatoire')
                .setColor(0x0099ff)
                .setTimestamp()
                .addFields({ name: '🔫 Votre arme', value: `\`\`\`${arme}\`\`\``, inline: false })
                .setFooter({ text: '🍀 Bonne chance !' });
            
            // Create back button to return to class view
            const armeRow = new ActionRowBuilder()
                .addComponents(
                    new ButtonBuilder()
                        .setCustomId('back_to_class')
                        .setLabel('🔙 Retour à la classe')
                        .setStyle(ButtonStyle.Secondary)
                );
            
            await interaction.editReply({ embeds: [armeEmbed], components: [armeRow] });
            break;

        case 'back_to_class':
            await interaction.deferUpdate();
            const classEmbed = createClassEmbed(classe);
            const classView = createRollView(classe);
            await interaction.editReply({ embeds: [classEmbed], ...classView });
            break;

        case 'defi':
            await interaction.deferUpdate();
            // Import defi view dynamically to avoid circular imports
            const { createDefiView, createDefiEmbed } = await import('./defiView.js');
            const defiEmbed = createDefiEmbed();
            const defiView = createDefiView(classe);
            await interaction.editReply({ embeds: [defiEmbed], ...defiView });
            break;

        case 'arme_principale_direct':
            updateStats('principale');
            const { createPrincipaleEmbed, createPrincipaleView } = await import('./principaleView.js');
            const principaleEmbed = createPrincipaleEmbed();
            const principaleView = createPrincipaleView();
            await interaction.reply({ embeds: [principaleEmbed], ...principaleView });
            break;

        case 'arme_secondaire_direct':
            updateStats('secondaire');
            const { createSecondaireEmbed, createSecondaireView } = await import('./secondaireView.js');
            const secondaireEmbed = createSecondaireEmbed();
            const secondaireView = createSecondaireView();
            await interaction.reply({ embeds: [secondaireEmbed], ...secondaireView });
            break;

        default:
            return null;
    }
    
    return null;
}